### PR TITLE
Telemetry:add app-client-configuration-change event

### PR DESCRIPTION
**Pull request by marcotc**
This PR adds the telemetry event `app-client-configuration-change`, which reports that configuration at the Client Library (`ddtrace`) has changed.
This event captures the changes themselves and a change source (called `origin` in the event).

The first use case is reporting changes from Dynamic Configuration. These have the origin `remote_config`.
The changes are a list of 2-tuples (`List<String, Integer | Boolean | String>`) representing the configuration that was just set.

One thing to noticed is that `app-client-configuration-change` is a Telemetry V2 event, but `ddtrace` currently implements Telemetry V1.
To aid with the migration to V2, the Datadog backend supports a few V2 events through the V1 pipeline. This is one of them.

This PR does not actually report the new event, only creates the API to trigger this event.

This event is crucial to Dynamic Configuration because it reports back to Datadog that the changes have been applied successfully to the current Ruby process.

### CODE REVIEW

File: lib/datadog/core/telemetry/client.rb [hunk 1]
```diff
@@ -65,6 +67,13 @@ def integrations_change!
           @emitter.request(:'app-integrations-change')
         end
 
+        # Report configuration changes caused by Dynamic Configuration.
+        def dynamic_configuration_change!(changes)
```
Comment (GustavoCaso) [inline lib/datadog/core/telemetry/client.rb:71]: Looks like there is a mismatch in the naming. `dynamic_configuration` and `remote_config`

File: spec/datadog/core/telemetry/emitter_spec.rb [hunk 2]
```diff
@@ -74,6 +74,17 @@
...
+
+    context 'with data' do
+      subject(:request) { emitter.request(request_type, data: data) }
+      let(:data) { { changes: ['test-data'] } }
+      let(:request_type) { 'app-client-configuration-change' }
+
+      it 'sends data to the emitter' do
+        request
+        expect(http_transport).to have_received(:request).with(payload: include('test-data'), request_type: anything)
+      end
```
Comment (GustavoCaso) [inline spec/datadog/core/telemetry/emitter_spec.rb:86]: Should this test focus on testing that `data` gets propagated to `Datadog::Core::Telemetry::Event.new.telemetry_request`? If feels that checking against `http_transport` is skipping the real change on `Emitter`

File: spec/datadog/core/telemetry/emitter_spec.rb [hunk 3]
```diff
@@ -74,6 +74,17 @@
         end
       end
     end
+
+    context 'with data' do
+      subject(:request) { emitter.request(request_type, data: data) }
+      let(:data) { { changes: ['test-data'] } }
+      let(:request_type) { 'app-client-configuration-change' }
+
+      it 'sends data to the emitter' do
```
Comment (GustavoCaso) [inline spec/datadog/core/telemetry/emitter_spec.rb:83]: The test description mentions emitter, but the asserting against `http_transport`.
Is that correct?

**Review by GustavoCaso — APPROVED** [review]
Left a few comments but overall LGTM

File: lib/datadog/core/telemetry/client.rb (hunk 1 above)
Comment (marcotc) [reply lib/datadog/core/telemetry/client.rb:71]: Renamed it to match the telemetry event.

File: spec/datadog/core/telemetry/emitter_spec.rb (hunk 2 above)
Comment (marcotc) [reply spec/datadog/core/telemetry/emitter_spec.rb:86]: changed it.

File: spec/datadog/core/telemetry/emitter_spec.rb (hunk 3 above)
Comment (marcotc) [reply spec/datadog/core/telemetry/emitter_spec.rb:83]: Renamed to match!

**Comment by marcotc** [comment]
Thank you, @GustavoCaso!

**Pull request merged**